### PR TITLE
Fix TypeError on `postToChat` in Python 3

### PR DESCRIPTION
**Pull request by doismellburning**
`TypeError: 'str' does not support the buffer interface`

https://docs.python.org/3/library/socket.html#socket.socket.sendall -
sendall takes bytes, and "foo" isn't bytes in Py3!

(This is a terrible bug report because I don't have internet access from the machine I demonstrated it on...)

### CODE REVIEW

**Comment by doismellburning** [comment]
Bah, that shouldn't have passed...

**Comment by doismellburning** [comment]
(I also don't have working Python 3 on the machine with internet access...)

**Comment by doismellburning** [comment]
Oh right, this won't work because we mocked out `mcpi.connection.Connection._send`

**Comment by doismellburning** [comment]
(Will merge in master once #11 is in)

File: tests/test_usage.py [hunk 1]
```diff
@@ -20,8 +21,17 @@
...
+        https://docs.python.org/3/library/socket.html#socket.socket.sendall
+
+        `sendall` takes _`bytes`_ explicitly in Python 3, _not_ `str`
+        """
+
+        assert isinstance(b, six.binary_type), "socket.socker.sendall was given non-bytes (%s)" % str(type(b))
+
     monkeypatch.setattr("socket.socket.connect", lambda x, y: None)
-    monkeypatch.setattr("socket.socket.sendall", lambda x, y: None)
+    monkeypatch.setattr("socket.socket.sendall", lambda x, y: dummy_sendall)
```
Comment (doismellburning) [inline tests/test_usage.py]: Yeah that's not what I meant, is it...

**Comment by doismellburning** [comment]
(Still won't fail until it has #11, but actually _does_ fail with #11 now!)

**Comment by ghickman** [comment]
@doismellburning – I've merged #11, is this now good to go?

**Comment by doismellburning** [comment]
Nope, just reproduces the bug - I haven't fixed it yet, I wanted to properly red-green things

**Comment by doismellburning** [comment]
Right. Fixing this is trivial: convert the string to bytes. Fixing it _nicely_ is harder.

It does not look like the server uses UTF-8:

```
>>> "ü".encode("utf-8")
b'\xc3\xbc'
>>> mc.conn._send(b"chat.post(hello\xc3\xbc)\n")
```

gives me `hello├╝` (those last two took some hunting to get as text - https://en.wikipedia.org/wiki/Code_page_437 to the rescue)

The protocol spec says "ascii".

I do not want to have to force users to write `mc.postToChat(b"Hello")`, otoh I do not want arbitrary decode errors

**Comment by doismellburning** [comment]
Right. `encode("cp437")`

Sadness. Sadness and woe. So much sadness and woe.

**Comment by doismellburning** [comment]
(thanks @lukasa and @mas90!)

**Comment by doismellburning** [comment]
I think we:

* Leave `Connection.send` as taking strings
* Leave `Connection._send` as taking strings

This means minimal refactoring of other code.

* Have `Connection._send` call `s.encode("cp437")` and try to produce a nice and helpful error message

I'm not sure how best to do this.

This is unfriendly:

```
>>> "☃".encode("cp437")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.3/lib/python3.3/encodings/cp437.py", line 12, in encode
    return codecs.charmap_encode(input,errors,encoding_map)
UnicodeEncodeError: 'charmap' codec can't encode character '\u2603' in position 0: character maps to <undefined>
```

I would like to provide a more helpful exception message explaining the `cp437` encoding, linking to the Wikipedia page with all the allowed characters etc., while NOT losing the original `UnicodeEncodeError`.

This seems hard to do generally, even without supporting all these Python versions

**Comment by doismellburning** [comment]
Actually, the lower we do the conversion, the less helpful the error will be :(

**Comment by doismellburning** [comment]
So:

* Make `Connection.send` take `bytes`, touch many many many lines of code to change `""`s into `b""`s, BUT have a nice friendly error message if you try to give non-CP437 strings to the few API methods that take strings

OR

* Tweak the internals of `Connection.send` to convert `string`s to `bytes`s and provide a bit of a generic message with included command data etc.?

**Comment by ghickman** [comment]
@doismellburning – my understanding of encoding is flaky at the best of times so I have one question for you:

What will be the nicer interface for users?

I think we should err on the side of making it nicer for them. Hope that helps! (this looks so painful)

**Comment by doismellburning** [comment]
@ghickman Good reminder, cheers. Nicest for users is:

> Make Connection.send take bytes, touch many many many lines of code to change ""s into b""s, BUT have a nice friendly error message if you try to give non-CP437 strings to the few API methods that take strings

I'll crack on with that, thanks

**Comment by doismellburning** [comment]
https://github.com/py3minepi/py3minepi/blob/fc60a2b5e8c243b4ee0c438d1cc61bd91fac156c/mcpi/util.py

This made us sad a while back.

It still causes sadness

**Comment by doismellburning** [comment]
* API preserved
* Test suite improved

BUT

* Not yet tested on Pi (doing now)
* Doesn't yet have nice hand-hold-y exception interface to non-CP437 encode-ables

I'd like to propose classifying the exception handling as "a new feature" (and thus out of scope for this), because, well, it is - the current behaviours pre- and post- bugfix are identical.

(And I don't want to block release on wrangling exceptions in a 2+3 way...)

**Comment by doismellburning** [comment]
(However, I would _love_ input on the refactor please - this felt very hard to document in a nice way)

**Comment by doismellburning** [comment]
Success!

**Comment by doismellburning** [comment]
photo 29-06-2015 22 35 03
photo 29-06-2015 22 34 49

:sunglasses:

**Comment by doismellburning** [comment]
No those two machines don't have interwebs

Nor are they mounted upside-down...

**Comment by doismellburning** [comment]
I did all of https://www.raspberrypi.org/learning/getting-started-with-minecraft-pi/worksheet/ down to the flowers bit, but figure you don't need more upside-down photos of my screens...

File: tests/test_usage.py [hunk 2]
```diff
@@ -20,8 +21,17 @@
...
 @pytest.fixture(autouse=True)
 def mc(monkeypatch):
+    def dummy_sendall(self, b):
+        """
+        https://docs.python.org/3/library/socket.html#socket.socket.sendall
+
+        `sendall` takes _`bytes`_ explicitly in Python 3, _not_ `str`
+        """
+
+        assert isinstance(b, six.binary_type), "socket.socker.sendall was given non-bytes (%s)" % str(type(b))
```
Comment (ghickman) [inline tests/test_usage.py]: `socker` seems wrong here?

**Comment by ghickman** [comment]
@doismellburning – this looks great. Assuming the `socker` message is wrong then this will be good to merge after that.

**Comment by doismellburning** [comment]
My love for Docker is sneaking into my messages about sockets. Thanks for spotting!

**Comment by doismellburning** [comment]
(Aaaaaaaaaargh!!!!)

**Pull request merged**